feat: add reflectivity to depth map output

seixasxbr opened the pull request.

seixasxbr committed:
feat: add reflectivity to depth map output

wvmcastro commented on tasks/Task.cpp:
```diff
@@ -156,13 +156,31 @@ void Task::convertDataAndWriteOutput(LidarScan& scan)
+    auto range = scan.field(sensor::ChanField::RANGE);
+    auto range_destaggered =
+        destagger<uint32_t>(range, m_metadata.format.pixel_shift_by_row);
+    Eigen::Array<uint32_t, -1, -1, Eigen::RowMajor> reflectivity;
+    if (m_metadata.format.udp_profile_lidar ==
+        sensor::UDPProfileLidar::PROFILE_LIDAR_LEGACY) {
+        reflectivity = scan.field(sensor::ChanField::REFLECTIVITY);
+    }
+    else if (m_metadata.format.udp_profile_lidar ==
+             sensor::UDPProfileLidar::PROFILE_RNG19_RFL8_SIG16_NIR16_DUAL) {
+        reflectivity =
+            scan.field<uint8_t>(sensor::ChanField::REFLECTIVITY).cast<uint32_t>();
+    }
+    else { // legacy or single return profile
+        reflectivity =
+            scan.field<uint16_t>(sensor::ChanField::REFLECTIVITY).cast<uint32_t>();
+    }
+    auto reflectivity_destaggered =
+        destagger<uint32_t>(reflectivity, m_metadata.format.pixel_shift_by_row);
```
`getReflectivity(LidarScan const&)`

wvmcastro commented on tasks/Task.cpp:
```diff
@@ -156,13 +156,31 @@ void Task::convertDataAndWriteOutput(LidarScan& scan)
     for (unsigned int h = 0; h < height; h++) {
         for (unsigned int w = 0; w < width; w++) {
-            depth_map.distances[h * width + w] = static_cast<double>(data(h, w)) / 1000.0;
+            depth_map.distances[h * width + w] = static_cast<double>(range_destaggered(h, w)) / 1000.0;
+            depth_map.remissions[h * width + w] = static_cast<double>(reflectivity_destaggered(h, w));
```
```suggestion
            const std::size_t index = h*width + w;
            depth_map.distances[index] = static_cast<double>(range_destaggered(h, w)) / 1000.0;
            depth_map.remissions[index] = static_cast<double>(reflectivity_destaggered(h, w));
```

wvmcastro reviewed (commented):
Agora o método `convertDataAndWriteOutput` está fazendo muitas coisas. Uma sugestão para quebrar ele em partes menores seria:
```cpp
convertDataAndWriteOutput(LidarScan& scan)
{
  base::Samples::DepthMap depth_map;
  fillMetadata(depth_map);
  auto reflectivity = getReflectivity(scan);
  ...
}
```
outra coisa que foi alterada nessa mudança é que a definição das variáveis `width` e `height` ficou distante do uso delas lá embaixo no `for`.

seixasxbr committed:
chore: get the reflectivity in a separete method

seixasxbr commented:
> Agora o método `convertDataAndWriteOutput` está fazendo muitas coisas. Uma sugestão para quebrar ele em partes menores seria:
> 
> ```c++
> convertDataAndWriteOutput(LidarScan& scan)
> {
>   base::Samples::DepthMap depth_map;
>   fillMetadata(depth_map);
>   auto reflectivity = getReflectivity(scan);
>   ...
> }
> ```
> 
> outra coisa que foi alterada nessa mudança é que a definição das variáveis `width` e `height` ficou distante do uso delas lá embaixo no `for`.

Na verdade o método só preenche os campos do depth map, acho que está bom assim. Mas adcionei o getReflectivity

seixasxbr committed:
chore: remove unecessary boolean

seixasxbr committed:
fix: remove reflectivity cast to uint32_t

since the reflectivity is always an array of 8 bits, it is uncessary to cast it to 32 bits

seixasxbr committed:
feat: enable compiling with optimization

seixasxbr committed:
fix: change buffer type to vector of uint8_t

seixasxbr committed:
fix: change method to get the imu data from buffer

use reinterpret_cast instead of memcpy

seixasxbr committed:
feat: add method to initialize the depth map in the start hook

depth map now is declared as a member class and the non-variable members of this type are defined in the start hook

seixasxbr committed:
fix: fix the remission calculation

The remission should be a value between 0 and 1. The LiDAR returns a reflectivity value between 0 and 255. Reflectivity values between 0-100 are linearly mapped for lambertian targets with values between 0% and 100% reflectivity. Values between 101-255 are mapped as log 2 with linear interpolation between logarithmic points for retroreflective targets. The 255 value corresponds to a retroreflector 864 times stronger than a 100% lambertian target. The charts below show the mapping functions.

wvmcastro commented on tasks/Task.cpp:
```diff
@@ -135,37 +149,55 @@ void Task::writeIMUSample(std::unique_ptr<uint8_t[]> const& pkt_buffer)
-            depth_map.distances[h * width + w] = static_cast<double>(data(h, w)) / 1000.0;
+            unsigned int index = h * width + w;
+            m_depth_map.distances[index] =
+                static_cast<double>(range_destaggered(h, w)) / 1000.0;
+            if (reflectivity_destaggered(h, w) <= 100) {
+                m_depth_map.remissions[index] =
+                    static_cast<double>(reflectivity_destaggered(h, w) / 100.0);
```
When possible, we should use multiplication over division. Especially in computer intensive tasks like this one.

source: https://stackoverflow.com/a/17883577

```suggestion
                static_cast<double>(range_destaggered(h, w)) * 1e-3;
            if (reflectivity_destaggered(h, w) <= 100) {
                m_depth_map.remissions[index] =
                    static_cast<double>(reflectivity_destaggered(h, w) * 1e-2);
```

seixasxbr replied:
good call

wvmcastro reviewed (approved):
Left just a tiny optimization obs

seixasxbr committed:
feat: add option to enable or disable remission

When the remission is enabled, the update frequency of the component drops to half. So, this modification allows to enable or disable the remission

seixasxbr committed:
fix: use multiplication instead of division to calculate remission

using multiplication in place of division could provide some optimization

seixasxbr committed:
fix: set standby mode in the cleanup hook

seixasxbr committed:
feat: add property to enable remission in test

seixasxbr committed:
chore: add 'using' directives to clear the code
